File: sdk/typescript/src/resources/iam.ts
import type { MashgateClient } from "../client.js";
import type {
  Permission,
  Role,
  Group,
  Policy,
  ApiKey,
  AuditEvent,
  Tenant,
  TenantQuota,
  AppScope,
} from "../types.js";

// ── Roles ──────────────────────────────────────────────────────────────────

export interface UpsertRoleRequest {
  tenantId: string;
  roleId?: string;
  code: string;
  name: string;
  permissions: string[];
  metadata?: Record<string, string>;
}

// ── Groups ─────────────────────────────────────────────────────────────────

export interface UpsertGroupRequest {
  tenantId: string;
  groupId?: string;
  code: string;
  name: string;
  metadata?: Record<string, string>;
}

// ── Policies ───────────────────────────────────────────────────────────────

export interface UpsertPolicyRequest {
  tenantId: string;
  policyId?: string;
  code: string;
  description?: string;
  conditionsJson: string;
  metadata?: Record<string, string>;
}

export interface BindPolicyRequest {
  tenantId: string;
  policyId: string;
  bindingType: string;
  bindingId: string;
}

// ── API Keys ───────────────────────────────────────────────────────────────

export interface CreateApiKeyRequest {
  tenantId: string;
  clientId: string;
  name: string;
  scopes?: string[];
  rpm?: number;
  burst?: number;
  expiresAt?: number;
  appId?: string;
}

export interface CreateApiKeyResponse {
  apiKey: ApiKey;
  plaintextKey: string;
}

export interface RotateApiKeyResponse {
  apiKey: ApiKey;
  plaintextKey: string;
}

// ── Access evaluation ──────────────────────────────────────────────────────

export interface EvaluateAccessRequest {
  tenantId: string;
  principalId: string;
  permission: string;
  resourceContext?: Record<string, string>;
}

export interface EvaluateAccessResponse {
  allow: boolean;
  reason: string;
  matchedPolicies: string[];
}

// ── Audit ──────────────────────────────────────────────────────────────────

export interface ListAuditEventsOptions {
  tenantId: string;
  page?: number;
  pageSize?: number;
}

export interface ListAuditEventsResponse {
  events: AuditEvent[];
  totalCount: number;
}

// ── App scopes ─────────────────────────────────────────────────────────────

export interface RegisterAppScopeRequest {
  clientId: string;
  scopeCode: string;
  description?: string;
}

export interface GrantScopeToRoleRequest {
  tenantId: string;
  roleId: string;
  scopeCode: string;
  clientId: string;
}

export interface GetEffectiveScopesRequest {
  tenantId: string;
  principalId: string;
  clientId: string;
}

// ── Tenants ────────────────────────────────────────────────────────────────

export interface CreateTenantRequest {
  /** Unique slug, e.g. "acme-corp". Stored as `code` in Mashgate. */
  code: string;
  name: string;
  /** "sandbox" | "live". Defaults to "sandbox". */
  mode?: string;
  /** Arbitrary key-value pairs. Use to link back to your system, e.g. { grid_workspace_id: "uuid" } */
  metadata?: Record<string, string>;
}

export interface CreateTenantResponse {
  tenantId: string;
  code: string;
  name: string;
  mode: string;
  /** "pending" | "active" | "failed" — provisioning is async */
  provStatus: string;
  createdAt: number;
}

export interface TenantProvisioningStatus {
  tenantId: string;
  /** "pending" | "active" | "failed" | "deleting" | "deleted" */
  provStatus: string;
  errorMsg?: string;
  attempt: number;
  updatedAt: number;
}

export interface ListTenantsOptions {
  page?: number;
  pageSize?: number;
  search?: string;
  status?: string;
  sortBy?: string;
  sortOrder?: string;
}

export interface ListTenantsResponse {
  tenants: Tenant[];
  totalCount: number;
}

export interface UpdateTenantRequest {
  name?: string;
  mode?: string;
  metadata?: Record<string, string>;
}

export interface SuspendTenantResponse {
  success: boolean;
  tenant: Tenant;
}

export interface BulkTenantActionRequest {
  tenantIds: string[];
  /** "suspend" | "activate" | "delete" */
  action: string;
}

export interface BulkTenantActionResponse {
  affectedCount: number;
  failedIds: string[];
}

// ── Resource class ─────────────────────────────────────────────────────────

export class IamResource {
  constructor(private readonly client: MashgateClient) {}

  // Permissions
  async listPermissions(tenantId: string): Promise<Permission[]> {
    const res = await this.client.request<{ permissions: Permission[] }>("GET", "/v1/iam/permissions", {
      query: { tenantId },
    });
    return res.permissions;
  }

  // Roles
  async listRoles(tenantId: string): Promise<Role[]> {
    const res = await this.client.request<{ roles: Role[] }>("GET", "/v1/iam/roles", {
      query: { tenantId },
    });
    return res.roles;
  }

  async upsertRole(data: UpsertRoleRequest): Promise<Role> {
    const res = await this.client.request<{ role: Role }>("POST", "/v1/iam/roles", { body: data });
    return res.role;
  }

  async deleteRole(tenantId: string, roleId: string): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>("DELETE", `/v1/iam/roles/${roleId}`, {
      query: { tenantId },
    });
    return res.success;
  }

  async assignRole(tenantId: string, principalId: string, roleId: string): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>("POST", "/v1/iam/roles/assign", {
      body: { tenantId, principalId, roleId },
    });
    return res.success;
  }

  async revokeRole(tenantId: string, principalId: string, roleId: string): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>("POST", "/v1/iam/roles/revoke", {
      body: { tenantId, principalId, roleId },
    });
    return res.success;
  }

  // Groups
  async listGroups(tenantId: string): Promise<Group[]> {
    const res = await this.client.request<{ groups: Group[] }>("GET", "/v1/iam/groups", {
      query: { tenantId },
    });
    return res.groups;
  }

  async upsertGroup(data: UpsertGroupRequest): Promise<Group> {
    const res = await this.client.request<{ group: Group }>("POST", "/v1/iam/groups", { body: data });
    return res.group;
  }

  async deleteGroup(tenantId: string, groupId: string): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>("DELETE", `/v1/iam/groups/${groupId}`, {
      query: { tenantId },
    });
    return res.success;
  }

  async addGroupMember(tenantId: string, groupId: string, principalId: string): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>("POST", `/v1/iam/groups/${groupId}/members`, {
      body: { tenantId, principalId },
    });
    return res.success;
  }

  async removeGroupMember(tenantId: string, groupId: string, principalId: string): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>(
      "DELETE",
      `/v1/iam/groups/${groupId}/members/${principalId}`,
      { query: { tenantId } },
    );
    return res.success;
  }

  // Policies
  async upsertPolicy(data: UpsertPolicyRequest): Promise<Policy> {
    const res = await this.client.request<{ policy: Policy }>("POST", "/v1/iam/policies", { body: data });
    return res.policy;
  }

  async deletePolicy(tenantId: string, policyId: string): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>("DELETE", `/v1/iam/policies/${policyId}`, {
      query: { tenantId },
    });
    return res.success;
  }

  async bindPolicy(data: BindPolicyRequest): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>("POST", "/v1/iam/policies/bind", { body: data });
    return res.success;
  }

  async unbindPolicy(data: BindPolicyRequest): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>("POST", "/v1/iam/policies/unbind", { body: data });
    return res.success;
  }

  // API Keys
  async listApiKeys(tenantId: string): Promise<ApiKey[]> {
    const res = await this.client.request<{ apiKeys: ApiKey[] }>("GET", "/v1/iam/api-keys", {
      query: { tenantId },
    });
    return res.apiKeys;
  }

  async createApiKey(data: CreateApiKeyRequest): Promise<CreateApiKeyResponse> {
    return this.client.request<CreateApiKeyResponse>("POST", "/v1/iam/api-keys", { body: data });
  }

  async rotateApiKey(tenantId: string, apiKeyId: string): Promise<RotateApiKeyResponse> {
    return this.client.request<RotateApiKeyResponse>("POST", `/v1/iam/api-keys/${apiKeyId}/rotate`, {
      body: { tenantId },
    });
  }

  async revokeApiKey(tenantId: string, apiKeyId: string): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>("DELETE", `/v1/iam/api-keys/${apiKeyId}`, {
      query: { tenantId },
    });
    return res.success;
  }

  // Access evaluation
  async evaluateAccess(data: EvaluateAccessRequest): Promise<EvaluateAccessResponse> {
    return this.client.request<EvaluateAccessResponse>("POST", "/v1/iam/evaluate", { body: data });
  }

  // Audit
  async listAuditEvents(options: ListAuditEventsOptions): Promise<ListAuditEventsResponse> {
    return this.client.request<ListAuditEventsResponse>("GET", "/v1/iam/audit-events", {
      query: {
        tenantId: options.tenantId,
        page: options.page,
        pageSize: options.pageSize,
      },
    });
  }

  // App scopes
  async registerAppScope(data: RegisterAppScopeRequest): Promise<AppScope> {
    const res = await this.client.request<{ scope: AppScope }>("POST", "/v1/iam/app-scopes", { body: data });
    return res.scope;
  }

  async listAppScopes(clientId: string): Promise<AppScope[]> {
    const res = await this.client.request<{ scopes: AppScope[] }>("GET", "/v1/iam/app-scopes", {
      query: { clientId },
    });
    return res.scopes;
  }

  async grantScopeToRole(data: GrantScopeToRoleRequest): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>("POST", "/v1/iam/app-scopes/grant", { body: data });
    return res.success;
  }

  async revokeScopeFromRole(tenantId: string, roleId: string, scopeCode: string): Promise<boolean> {
    const res = await this.client.request<{ success: boolean }>("POST", "/v1/iam/app-scopes/revoke", {
      body: { tenantId, roleId, scopeCode },
    });
    return res.success;
  }

  async getEffectiveScopes(data: GetEffectiveScopesRequest): Promise<string[]> {
    const res = await this.client.request<{ scopes: string[] }>("GET", "/v1/iam/effective-scopes", {
      query: { tenantId: data.tenantId, principalId: data.principalId, clientId: data.clientId },
    });
    return res.scopes;
  }

  // Tenant lifecycle
  async createTenant(data: CreateTenantRequest): Promise<CreateTenantResponse> {
    return this.client.request<CreateTenantResponse>("POST", "/v1/iam/tenants", { body: data });
  }

  async getTenantProvisioningStatus(tenantId: string): Promise<TenantProvisioningStatus> {
    return this.client.request<TenantProvisioningStatus>(
      "GET",
      `/v1/iam/tenants/${tenantId}/provisioning`,
    );
  }

  async deleteTenant(tenantId: string): Promise<{ accepted: boolean; provStatus: string }> {
    return this.client.request("DELETE", `/v1/iam/tenants/${tenantId}`);
  }

  // Admin tenant management
  async listTenants(options?: ListTenantsOptions): Promise<ListTenantsResponse> {
    return this.client.request<ListTenantsResponse>("GET", "/v1/iam/tenants", { query: options as any });
  }

  async getTenant(tenantId: string): Promise<Tenant> {
    return this.client.request<Tenant>("GET", `/v1/iam/tenants/${tenantId}`);
  }

  async updateTenant(tenantId: string, data: UpdateTenantRequest): Promise<Tenant> {
    return this.client.request<Tenant>("PUT", `/v1/iam/tenants/${tenantId}`, { body: data });
  }

  async suspendTenant(tenantId: string, reason?: string): Promise<SuspendTenantResponse> {
    return this.client.request<SuspendTenantResponse>("POST", `/v1/iam/tenants/${tenantId}/suspend`, {
      body: { reason },
    });
  }

  async listTenantQuotas(tenantId: string): Promise<TenantQuota[]> {
    const res = await this.client.request<{ quotas: TenantQuota[] }>(
      "GET",
      `/v1/iam/tenants/${tenantId}/quotas`,
    );
    return res.quotas;
  }

  async updateTenantQuota(tenantId: string, resource: string, limit: number): Promise<TenantQuota> {
    return this.client.request<TenantQuota>(
      "PUT",
      `/v1/iam/tenants/${tenantId}/quotas/${resource}`,
      { body: { limit } },
    );
  }

  async bulkTenantAction(data: BulkTenantActionRequest): Promise<BulkTenantActionResponse> {
    return this.client.request<BulkTenantActionResponse>("POST", "/v1/iam/tenants/bulk", { body: data });
  }
}

File: sdk/python/mashgate/resources/iam.py
"""IAM resource (mgID).

Mirrors ``iam.v1.IamService`` exposed over the gateway as REST.

Surface unions the Go SDK (``sdk/go/iam.go``) and the TypeScript SDK
(``sdk/typescript/src/resources/iam.ts``):

- Tenant identity + lifecycle (create / get / list / update / suspend /
  delete / provisioning-status / quotas / bulk actions).
- Roles, groups, policies (RBAC + ABAC).
- API keys (create / list / rotate / revoke).
- Access evaluation (``evaluate`` ABAC, ``check`` simple permission probe).
- Audit events and application scopes.

All methods return the parsed JSON ``dict`` from the gateway. Callers that
want a specific sub-field (e.g. ``["roles"]``) pull it from the returned
dict; we keep the full envelope so callers also see pagination / counts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from mashgate.client import MashgateClient


class IamResource:
    """Client for the full ``iam.v1.IamService`` (mgID)."""

    def __init__(self, client: MashgateClient) -> None:
        self._c = client

    # ── Permissions ───────────────────────────────────────────────────

    def list_permissions(self, tenant_id: str) -> dict[str, Any]:
        return self._c.request(
            "GET", "/v1/iam/permissions", query={"tenantId": tenant_id}
        )

    # ── Roles ─────────────────────────────────────────────────────────

    def list_roles(self, tenant_id: str) -> dict[str, Any]:
        return self._c.request("GET", "/v1/iam/roles", query={"tenantId": tenant_id})

    def upsert_role(
        self,
        *,
        tenant_id: str,
        code: str,
        name: str,
        permissions: list[str],
        role_id: str | None = None,
        metadata: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "tenantId": tenant_id,
            "code": code,
            "name": name,
            "permissions": permissions,
        }
        if role_id is not None:
            body["roleId"] = role_id
        if metadata is not None:
            body["metadata"] = metadata
        return self._c.request("POST", "/v1/iam/roles", body=body)

    def delete_role(self, tenant_id: str, role_id: str) -> dict[str, Any]:
        return self._c.request(
            "DELETE", f"/v1/iam/roles/{role_id}", query={"tenantId": tenant_id}
        )

    def assign_role(
        self, tenant_id: str, principal_id: str, role_id: str
    ) -> dict[str, Any]:
        return self._c.request(
            "POST",
            "/v1/iam/roles/assign",
            body={
                "tenantId": tenant_id,
                "principalId": principal_id,
                "roleId": role_id,
            },
        )

    def revoke_role(
        self, tenant_id: str, principal_id: str, role_id: str
    ) -> dict[str, Any]:
        return self._c.request(
            "POST",
            "/v1/iam/roles/revoke",
            body={
                "tenantId": tenant_id,
                "principalId": principal_id,
                "roleId": role_id,
            },
        )

    # ── Groups ────────────────────────────────────────────────────────

    def list_groups(self, tenant_id: str) -> dict[str, Any]:
        return self._c.request("GET", "/v1/iam/groups", query={"tenantId": tenant_id})

    def upsert_group(
        self,
        *,
        tenant_id: str,
        code: str,
        name: str,
        group_id: str | None = None,
        metadata: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"tenantId": tenant_id, "code": code, "name": name}
        if group_id is not None:
            body["groupId"] = group_id
        if metadata is not None:
            body["metadata"] = metadata
        return self._c.request("POST", "/v1/iam/groups", body=body)

    def delete_group(self, tenant_id: str, group_id: str) -> dict[str, Any]:
        return self._c.request(
            "DELETE", f"/v1/iam/groups/{group_id}", query={"tenantId": tenant_id}
        )

    def add_group_member(
        self, tenant_id: str, group_id: str, principal_id: str
    ) -> dict[str, Any]:
        return self._c.request(
            "POST",
            f"/v1/iam/groups/{group_id}/members",
            body={"tenantId": tenant_id, "principalId": principal_id},
        )

    def remove_group_member(
        self, tenant_id: str, group_id: str, principal_id: str
    ) -> dict[str, Any]:
        return self._c.request(
            "DELETE",
            f"/v1/iam/groups/{group_id}/members/{principal_id}",
            query={"tenantId": tenant_id},
        )

    # ── Policies ──────────────────────────────────────────────────────

    def upsert_policy(
        self,
        *,
        tenant_id: str,
        code: str,
        conditions_json: str,
        policy_id: str | None = None,
        description: str | None = None,
        metadata: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {
            "tenantId": tenant_id,
            "code": code,
            "conditionsJson": conditions_json,
        }
        if policy_id is not None:
            body["policyId"] = policy_id
        if description is not None:
            body["description"] = description
        if metadata is not None:
            body["metadata"] = metadata
        return self._c.request("POST", "/v1/iam/policies", body=body)

    def delete_policy(self, tenant_id: str, policy_id: str) -> dict[str, Any]:
        return self._c.request(
            "DELETE", f"/v1/iam/policies/{policy_id}", query={"tenantId": tenant_id}
        )

    def bind_policy(
        self,
        *,
        tenant_id: str,
        policy_id: str,
        binding_type: str,
        binding_id: str,
    ) -> dict[str, Any]:
        return self._c.request(
            "POST",
            "/v1/iam/policies/bind",
            body={
                "tenantId": tenant_id,
                "policyId": policy_id,
                "bindingType": binding_type,
                "bindingId": binding_id,
            },
        )

    def unbind_policy(
        self,
        *,
        tenant_id: str,
        policy_id: str,
        binding_type: str,
        binding_id: str,
    ) -> dict[str, Any]:
        return self._c.request(
            "POST",
            "/v1/iam/policies/unbind",
            body={
                "tenantId": tenant_id,
                "policyId": policy_id,
                "bindingType": binding_type,
                "bindingId": binding_id,
            },
        )

    # ── API keys ──────────────────────────────────────────────────────

    def list_api_keys(self, tenant_id: str) -> dict[str, Any]:
        return self._c.request(
            "GET", "/v1/iam/api-keys", query={"tenantId": tenant_id}
        )

    def create_api_key(
        self,
        *,
        tenant_id: str,
        client_id: str,
        name: str,
        scopes: list[str] | None = None,
        rpm: int | None = None,
        burst: int | None = None,
        expires_at: int | None = None,
        app_id: str | None = None,
    ) -> dict[str, Any]:
        """Create a new API key.

        The returned ``plaintextKey`` is shown once — store it securely.
        """
        body: dict[str, Any] = {
            "tenantId": tenant_id,
            "clientId": client_id,
            "name": name,
        }
        if scopes is not None:
            body["scopes"] = scopes
        if rpm is not None:
            body["rpm"] = rpm
        if burst is not None:
            body["burst"] = burst
        if expires_at is not None:
            body["expiresAt"] = expires_at
        if app_id is not None:
            body["appId"] = app_id
        return self._c.request("POST", "/v1/iam/api-keys", body=body)

    def rotate_api_key(self, tenant_id: str, api_key_id: str) -> dict[str, Any]:
        """Rotate an API key. The returned ``plaintextKey`` is shown once."""
        return self._c.request(
            "POST",
            f"/v1/iam/api-keys/{api_key_id}/rotate",
            body={"tenantId": tenant_id},
        )

    def revoke_api_key(self, tenant_id: str, api_key_id: str) -> dict[str, Any]:
        return self._c.request(
            "DELETE",
            f"/v1/iam/api-keys/{api_key_id}",
            query={"tenantId": tenant_id},
        )

    # ── Access evaluation ─────────────────────────────────────────────

    def evaluate_access(
        self,
        *,
        tenant_id: str,
        principal_id: str,
        permission: str,
        resource_context: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Full ABAC evaluation. Returns ``{allow, reason, matchedPolicies}``."""
        body: dict[str, Any] = {
            "tenantId": tenant_id,
            "principalId": principal_id,
            "permission": permission,
        }
        if resource_context is not None:
            body["resourceContext"] = resource_context
        return self._c.request("POST", "/v1/iam/evaluate", body=body)

    def check_permission(self, permission: str) -> dict[str, Any]:
        """Probe whether the authenticated principal has ``permission``.

        Mirrors the Go SDK's ``CheckPermission``; hits
        ``GET /v1/iam/check?permission=...`` and returns ``{"allowed": bool}``.
        Useful as lightweight middleware: ``if not res["allowed"]: 403``.
        """
        return self._c.request(
            "GET", "/v1/iam/check", query={"permission": permission}
        )

    # ── Audit ─────────────────────────────────────────────────────────

    def list_audit_events(
        self,
        *,
        tenant_id: str,
        page: int | None = None,
        page_size: int | None = None,
    ) -> dict[str, Any]:
        return self._c.request(
            "GET",
            "/v1/iam/audit-events",
            query={"tenantId": tenant_id, "page": page, "pageSize": page_size},
        )

    # ── App scopes ────────────────────────────────────────────────────

    def register_app_scope(
        self, *, client_id: str, scope_code: str, description: str | None = None
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"clientId": client_id, "scopeCode": scope_code}
        if description is not None:
            body["description"] = description
        return self._c.request("POST", "/v1/iam/app-scopes", body=body)

    def list_app_scopes(self, client_id: str) -> dict[str, Any]:
        return self._c.request(
            "GET", "/v1/iam/app-scopes", query={"clientId": client_id}
        )

    def grant_scope_to_role(
        self, *, tenant_id: str, role_id: str, scope_code: str, client_id: str
    ) -> dict[str, Any]:
        return self._c.request(
            "POST",
            "/v1/iam/app-scopes/grant",
            body={
                "tenantId": tenant_id,
                "roleId": role_id,
                "scopeCode": scope_code,
                "clientId": client_id,
            },
        )

    def revoke_scope_from_role(
        self, tenant_id: str, role_id: str, scope_code: str
    ) -> dict[str, Any]:
        return self._c.request(
            "POST",
            "/v1/iam/app-scopes/revoke",
            body={
                "tenantId": tenant_id,
                "roleId": role_id,
                "scopeCode": scope_code,
            },
        )

    def get_effective_scopes(
        self, *, tenant_id: str, principal_id: str, client_id: str
    ) -> dict[str, Any]:
        return self._c.request(
            "GET",
            "/v1/iam/effective-scopes",
            query={
                "tenantId": tenant_id,
                "principalId": principal_id,
                "clientId": client_id,
            },
        )

    # ── Tenant lifecycle ──────────────────────────────────────────────

    def create_tenant(
        self,
        *,
        code: str,
        name: str,
        mode: str | None = None,
        metadata: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Create a Mashgate tenant. Provisioning is async — poll
        :meth:`get_tenant_provisioning_status` for completion.

        :param code: Unique slug, e.g. ``"acme-corp"``. Stored as ``code``.
        :param mode: ``"sandbox"`` | ``"live"``. Defaults to ``"sandbox"``.
        :param metadata: Arbitrary key-value pairs to link back to your
            system, e.g. ``{"grid_workspace_id": "uuid"}``.
        """
        body: dict[str, Any] = {"code": code, "name": name}
        if mode is not None:
            body["mode"] = mode
        if metadata is not None:
            body["metadata"] = metadata
        return self._c.request("POST", "/v1/iam/tenants", body=body)

    def get_tenant_provisioning_status(self, tenant_id: str) -> dict[str, Any]:
        return self._c.request(
            "GET", f"/v1/iam/tenants/{tenant_id}/provisioning"
        )

    def delete_tenant(self, tenant_id: str) -> dict[str, Any]:
        return self._c.request("DELETE", f"/v1/iam/tenants/{tenant_id}")

    # ── Admin tenant management ───────────────────────────────────────

    def list_tenants(
        self,
        *,
        page: int | None = None,
        page_size: int | None = None,
        search: str | None = None,
        status: str | None = None,
        sort_by: str | None = None,
        sort_order: str | None = None,
    ) -> dict[str, Any]:
        """List tenants visible to the authenticated principal.

        Used for cold-start backfill in downstream verticals before
        subscribing to the Kafka ``tenant-events`` topic (ADR-0020 Phase B).
        """
        return self._c.request(
            "GET",
            "/v1/iam/tenants",
            query={
                "page": page,
                "pageSize": page_size,
                "search": search,
                "status": status,
                "sortBy": sort_by,
                "sortOrder": sort_order,
            },
        )

    def get_tenant(self, tenant_id: str) -> dict[str, Any]:
        return self._c.request("GET", f"/v1/iam/tenants/{tenant_id}")

    def update_tenant(
        self,
        tenant_id: str,
        *,
        name: str | None = None,
        mode: str | None = None,
        metadata: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if mode is not None:
            body["mode"] = mode
        if metadata is not None:
            body["metadata"] = metadata
        return self._c.request("PUT", f"/v1/iam/tenants/{tenant_id}", body=body)

    def suspend_tenant(
        self, tenant_id: str, *, reason: str | None = None
    ) -> dict[str, Any]:
        body: dict[str, Any] = {}
        if reason is not None:
            body["reason"] = reason
        return self._c.request(
            "POST", f"/v1/iam/tenants/{tenant_id}/suspend", body=body or None
        )

    def list_tenant_quotas(self, tenant_id: str) -> dict[str, Any]:
        return self._c.request("GET", f"/v1/iam/tenants/{tenant_id}/quotas")

    def update_tenant_quota(
        self, tenant_id: str, resource: str, limit: int
    ) -> dict[str, Any]:
        return self._c.request(
            "PUT",
            f"/v1/iam/tenants/{tenant_id}/quotas/{resource}",
            body={"limit": limit},
        )

    def bulk_tenant_action(
        self, *, tenant_ids: list[str], action: str
    ) -> dict[str, Any]:
        """Apply ``action`` to many tenants at once.

        :param action: ``"suspend"`` | ``"activate"`` | ``"delete"``.
        :returns: ``{"affectedCount": int, "failedIds": [...]}``.
        """
        return self._c.request(
            "POST",
            "/v1/iam/tenants/bulk",
            body={"tenantIds": tenant_ids, "action": action},
        )
